pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -0,0 +1,114 @@
+package nstemplateset
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+
+	toolchainv1alpha1 "github.com/codeready-toolchain/api/pkg/apis/toolchain/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	projectv1 "github.com/openshift/api/project/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_nstemplateset")
+
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileNSTemplateSet{client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("nstemplateset-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource
+	err = c.Watch(&source.Kind{Type: &toolchainv1alpha1.NSTemplateSet{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileNSTemplateSet{}
+
+type ReconcileNSTemplateSet struct {
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a NSTemplateSet object and makes changes based on the state read
+// and what is in the NSTemplateSet.Spec
+func (r *ReconcileNSTemplateSet) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling NSTemplateSet")
+
+	// Fetch the NSTemplateSet instance
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, nsTmplSet)
+	if err != nil {
+		if errors.IsNotFound(err) {
```

Comment:
We will need to add create a JIRA for garbage collection of NSTemplateSet. Setting owner references won't help since we are introducing namespaced (templateSet) <-> cluster scoped (namespaces, project) relation here. So we will need to handle garbage collection manually.
Can you please create an issue for that?

Reply:
https://jira.coreos.com/browse/CRT-224